flight/mcu/lib/teensyShared/kalmanFilter.cpp

```diff
@@ -0,0 +1,167 @@
+     * estimate. This process is significantly faster than allowing the state as
+     * letting the filter to converge to the correct state can take up to 3 min.
+     * This specific process was used because the barometric altitude will
+     * change depending on the weather and thus, the initial state estimate
+     * cannot be hard coded. A GPS altitude may be used instead but due to GPS
+     * losses during high speed/high altitude flight, it is inadvisable with the
+     * current hardware to use this as a solution. Reference frames should also
+     * be kept consistent (do not mix GPS altitude and barometric).
+     */
+    float sum = 0;
+    for (int i = 0; i < 30; i++) {
+        chMtxLock(dataMutex_barometer_);
+        sum += *b_alt;
+        chMtxUnlock(dataMutex_barometer_);
+        chThdSleepMilliseconds(100);
+    }
+
+    // set x_k
+    x_k(0, 0) = sum / 30;
+    // x_k(0,0) = 1401;
+    x_k(1, 0) = 0;
+    x_k(2, 0) = 0;
+
+    // set F
+    F_mat(0, 1) = s_dt;
+    F_mat(0, 2) = (s_dt * s_dt) / 2;
+    F_mat(1, 2) = s_dt;
+
+    F_mat(0, 0) = 1;
+    F_mat(1, 1) = 1;
+    F_mat(2, 2) = 1;
+
+    // set H
+    H(0, 0) = 1;
+    H(1, 2) = 1;
+
+    // set P_k
+    P_k(0, 0) = 0;
+    P_k(0, 1) = 0;
+    P_k(0, 2) = 0;
+    P_k(1, 1) = 0;
+    P_k(2, 2) = 0;
+    P_k(1, 2) = 0;
+    P_k(2, 1) = P_k(1, 2);
+    P_k(1, 0) = P_k(0, 1);
+    P_k(2, 0) = P_k(0, 2);
+    P_k(1, 0) = P_k(0, 1);
+
+    // set Q
+    Q(0, 0) = pow(s_dt, 5) / 20;
+    Q(0, 1) = (pow(s_dt, 4) / 8 * 80);
+    Q(0, 2) = pow(s_dt, 3) / 6;
+    Q(1, 1) = pow(s_dt, 3) / 8;
+    Q(1, 2) = pow(s_dt, 2) / 2;
+    Q(2, 2) = s_dt;
+    Q(1, 0) = Q(0, 1);
+    Q(2, 0) = Q(0, 2);
+    Q(2, 1) = Q(1, 2);
+
+    // float scale_fact = 75.19;
+    // float scale_fact = 14.25;
+    float scale_fact = 13.0;
+    // float scale_fact = 13;
+    Q = Q * scale_fact;
+
+    // set R
+    R(0, 0) = 2.0;
+    R(1, 1) = 0.01;
+    // R(0,0) = 2.;
+    // R(1,1) = .01;
+
+    // set B
+    B(2, 0) = -1;
+}
+
+void KalmanFilter::Initialize(float pos_f, float vel_f) {
+    // set x_k
+    x_k(0, 0) = pos_f;
+    x_k(1, 0) = vel_f;
+
+    // set F
+    F_mat(0, 1) = s_dt;
+
+    F_mat(0, 0) = 1;
+    F_mat(1, 1) = 1;
+
+    // set H
+    H(0, 0) = 1;
+
+    // set R
+    R(0, 0) = 12;
+
+    // set B
+    B(2, 0) = -1;
+}
+
+void KalmanFilter::priori() {
+    // x_priori = (F @ x_k) + ((B @ u).T) #* For some reason doesnt work when B
+    // or u is = 0
+    x_priori = (F_mat * x_k);
+    P_priori = (F_mat * P_k * F_mat.transpose()) + Q;
+}
+
+void KalmanFilter::update() {
+    // Update Kalman Gain
+
+    Eigen::Matrix<float, 2, 2> temp = Eigen::Matrix<float, 2, 2>::Zero();
+    temp = (((H * P_priori * H.transpose()) + R)).inverse();
+    Eigen::Matrix<float, 3, 3> identity =
+        Eigen::Matrix<float, 3, 3>::Identity();
+    K = (P_priori * H.transpose()) * temp;
+
+    // Sensor Measurements
+    chMtxLock(mutex_highG_);
+    y_k(1, 0) = (*gz_H) * 9.81;
```

Comment:
Does this account for the fact that the High-G barometer gives specific force (is that the right term?) instead of true acceleration? It's not just a conversion of g's to m/s^2. Or am I mistaken @parthshrotri ?

Reply:
the high-g _IMU_ (we all need sleep lmao) outputs in g's as far as i know

Reply:
Yes it does, tested in silsim